core/runtime/src/main/java/io/quarkus/runtime/CracRecorder.java

```diff
@@ -0,0 +1,95 @@
+package io.quarkus.runtime;
+
+import java.io.BufferedReader;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.URL;
+import java.net.URLConnection;
+import java.util.Enumeration;
+
+import org.crac.Context;
+import org.crac.Resource;
+import org.jboss.logging.Logger;
+
+import io.quarkus.runtime.annotations.Recorder;
+
+/**
+ * Registers a CRAC resource. Must be called in static initialization phase!
+ */
+@Recorder
+public class CracRecorder {
+
+    protected static class CracResource implements Resource {
+        protected final boolean fullWarmup;
+        protected final boolean preloadClasses;
+        protected int numPreloaded;
+
+        public CracResource(boolean preloadClasses, boolean fullWarmup) {
+            this.preloadClasses = preloadClasses;
+            this.fullWarmup = fullWarmup;
+        }
+
+        @Override
+        public void beforeCheckpoint(Context<? extends Resource> context) throws Exception {
+            if (preloadClasses) {
+                preloadClasses();
+            }
+            if (fullWarmup) {
+                Quarkus.manualStart();
+                System.out.println("Started Quarkus via CRAC beforeCheckpoint()");
+            }
+        }
+
+        @Override
+        public void afterRestore(Context<? extends Resource> context) throws Exception {
+            if (!fullWarmup) {
+                Quarkus.manualStart();
+                Logger.getLogger(Quarkus.class).info("Started Quarkus via CRAC afterRestore()");
+            } else {
+                Logger.getLogger(Quarkus.class).info("Started Quarkus via CRAC beforeCheckpoint()");
+            }
+            if (preloadClasses)
+                Logger.getLogger(Quarkus.class).info("Number preloaded classes: " + numPreloaded);
+        }
+
+        protected void preloadClass(String classname) {
+            try {
+                numPreloaded++;
+                Class.forName(classname);
+            } catch (Throwable ignored) {
+
+            }
+        }
+
+        protected void preloadClasses() throws Exception {
```

Comment:
PR #29016 covers initialization (not just preloading) generically using build items rather than resources (until the back end) and is not specific to crac or even necessarily lambda, since preinitializing could be beneficial generally speaking. WDYT about that approach? Where is the perf gap mainly, on load or init?

Reply:
I'll comment on the PR, I have concerns about that approach.

Reply:
I'll comment here too.  Processing these preloaded-classes.txt files is unnecessary overhead.  Extensions and libraries can provide preloaded-classes.txt files in their META-INF at now cost, and only have these files processed in the environments that need them (i.e. CRAC).

Reply:
Both approaches are pros and cons. 
@dmlloyd's is more "quarky", @patriot1burke's has the benefits to avoid pre-loeading everywhere.

I think the best would be to mix both solutions.
What about:

1. a built-time config enabling / disabling the pre-loading (marked as experimental, disable by default)
2. we keep the build item from David's PR
3.in the processor, we check the config, if present we use a recorder to receive the list of classes and a public static method to preload
4. in the manual start, if the feature is enabled, we call the public static method to preload the classes

This approach has the advantage to integrate well with Quarkus (build item, recorder), while only using this feature if enabled explicitly (I share @patriot1burke's concerns about that feature in regular JVM app (it can have advantages, but also disadvantages)). Later, we could decide how we can detect if that flag must be enabled or disabled. 

WDYT?

Reply:
With DML's solution there is one preload file generated by the set of builditems.  While great for extensions issues that require one-off preload of a certain class, not great for bulk preloads where we may be loading hundreds, maybe thousands of classes.  Again...I envision each extension manually, by hand, providing these huge lists of classes to preload...Or even a user providing such a file too by running -verbose:class, scraping the results, and adding the preload file to their app's Jar.